Proposal: docker ps: add fields for ordering and column selection

codemac opened the pull request:
- api/client/commands.go: Refactor CmdPs to use a fields list of
  characters to determine which columns to print on `docker ps`
  invocation.

This adds an ability for the docker command to print the columns of
output in arbitrary order.

This is nominally related to #7477 

Signed-off-by: Jeff Mickey j@codemac.net

codemac committed:
docker ps: add fields for ordering and column selection

* api/client/ps.go: Refactor CmdPs to use a fields list of
  characters to determine which columns to print on `docker ps`
  invocation.

This adds an ability for the docker command to print the columns of
output in arbitrary order.

Signed-off-by: Jeff Mickey <j@codemac.net>

codemac commented:
This is my first contribution to the docker project, I apologize for any errors.

This adds a `-fields` to docker ps that allows you to choose which, and in what order the columns are printed. Haven't finished testing locally yet, but wanted to see if this is the completely wrong direction. This code is small and simple, so any feedback including rewrites is appreciated.

jfrazelle commented:
Thanks for your first contribution! We are super excited to have you! Now since your PR is changing the UX of docker ps it is going to need UX review. Seeing as we are in the middle of the release process, about to make release candidate binaries and try and test and fix all bugs that are found it might take a little longer for us to get to this PR. But we promise we are here and this seems really awesome. Just don't want you to be concerned if there is a slight lull in review :)

codemac commented:
Yea, more than anything the type of feedback I'd like is if this is the approach people agree with.

The amount of `docker ps | tail -n +2 | awk '{print $NF}'` occurring in my life is too damn high :)

codemac commented:
Please let me know what thoughts are, I can rebase/update if necessary.

SvenDowideit commented:
oh +1 - I was just wanting this for `docker image` too

codemac commented:
This is my monthly ping!

Also - the windows bulider failed pretty fantastically, on all kinds of things that don't look like my fault.

LK4D4 commented:
Code is little unreadable with all this `next` stuff, but idea is okay.
@tiborvass should we move to code-review?

tiborvass commented:
@LK4D4 although I understand the idea and the reason, I'm not 100% sure we want this (not saying I'm sure we don't want this :P )

Ping @crosbymichael @jfrazelle @estesp @duglin @icecrime

icecrime commented:
I *personally* don't feel a need for this, so I'm a "weak no".

thaJeztah commented:
Different proposal, but just for showing (selected) labels in `docker ps` https://github.com/docker/docker/pull/11943 might be good to compare the use-cases and see if they make sense together?

duglin commented on api/client/commands.go:
```diff
@@ -1656,13 +1657,35 @@ func (cli *DockerCli) CmdPs(args ...string) error {
+	}
+
+	if *size {
+		*fields = *fields + "z"
```
do you need to make sure 'z' isn't already in there?

duglin replied:
actually, do you purposely want to allow `--fields=cccc`  ?  Guess if that's what the user wants.....

codemac replied:
It was my consideration that it could be desired, but doesn't seem in line with what I think a user would suspect. I'll uniq the letters that are set in *fields, choosing the earliest existence.

duglin commented on api/client/commands.go:
```diff
@@ -1674,60 +1697,86 @@ func (cli *DockerCli) CmdPs(args ...string) error {
 
+	outp := make([]containerMeta, 0)
+	for _, out := range outs.Data {
+		portTable := engine.NewTable("", 0)
```
We're in the process of removing engine.Table stuff so you'll need a different way to do this.

codemac replied:
I'm not sure I know exactly what engine.Table stuff really does. Any pointers on what other commands are doing that depend on the table stuff? I could come up with a strawman but I don't know what y'all want.

duglin replied:
if you wait a bit (hopefully later today) and then rebase you will see what we're using instead.
I'll try to remember to let to know when to do it.

duglin replied:
ha - should have finished going thru my mail. Rebase against master and you should see the new stuff.

duglin commented on api/client/commands.go:
```diff
@@ -1674,60 +1697,86 @@ func (cli *DockerCli) CmdPs(args ...string) error {
-			} else {
-				fmt.Fprintf(w, "%s\n", units.HumanSize(float64(out.GetInt64("SizeRw"))))
-			}
+	for _, out := range outp {
```
Is there any reason not to print it immediately instead of stashing it into 'outp' ?

codemac replied:
Just so down below, I can do the simple fprintf + strings.Join ("\t").. otherwise I have to track if I'm at the last letter, add prints of \t outside the switch, etc. Looked a lot less nicer. Considering this is per line, the memory overhead seemed minimal.

duglin commented:
I actually like the idea and wouldn't mind seeing it on other tabular-output commands.
Major re-base is needed though - sorry - lots of changes these days.

duglin commented:
I guess before I make more comments on the code we need to exit design review :-)
I'm +1 on it.  @tiborvass  ?

thaJeztah commented:
I wonder if we only want the shorthand (single-chars) to specify the columns, I can imagine they will be hard to remember without checking the help/man. Perhaps also `--fields=id,image,status`

Would using a comma-separated list be good idea in general (also for single-chars), as it leaves more room for future expansion (I think)

tiborvass commented:
Still not sure if I want this. But if I did, I agree with @thaJeztah on separating actual words with commas. I don't like the single-chars.

Even though unix philosophy would say no to this PR, I agree it's not as easy to awk the output. One could say just use the API though. Maybe there's room for this. I just hope we're not opening a can of worms where people can point to this PR saying "but he got to have his --fields PR merged".

duglin commented:
@tiborvass can you elaborate on why on you think the unix philosophy would say no to this PR?  The linx 'ps' command itself has this feature.

cpuguy83 commented:
How about just `docker ps --json` to spit out the literally json data instead of building the table?

thaJeztah commented:
>  How about just docker ps --json to spit out the literally json data instead of building the table?

I think that option was opted somewhere before (in combination with a `--format` option). On one hand, I think it makes sense (after all, `docker inspect` also outputs JSON), on the other hand, it might become a slippery slope; when to use the API, and when the CLI (if both use a JSON interface)

aluzzardi commented:
I'd like to see this and #11943 merged together.

It's useful to only display certain columns, but since the introduction of labels it's even more useful to display custom labels alongside.

/cc @vieux

icecrime commented:
We had a collective review session today about this PR: we decided to move forward with it (as in: we want this feature). However, we might refine the UX in some ways: @calavera will take the lead on this and either work with @codemac to update the content, or carry the PR depending on the original author's availability. Thanks!

codemac commented:
* Arnaud Porterie <notifications@github.com> [2015-04-30 22:28]:
> @calavera will take the lead on this and either work with @codemac to
> update the content, or carry the PR depending on the original author's
> availability.

I'm very available to make this change, I code in go all day and use
docker every day :)

Please let me know what UX changes are required -- I think most of the
suggestions (full field names vs letters, comma separation vs not,
output formats in general) were all very implementable.

I like the ps-style single letter thing best, but the thrust of the pr
is the ability to choose columns / orders of text for unix-y shell
goodness.

  //  codemac

aluzzardi commented:
/cc @vieux @abronan

If this we merged, we'll need to port it to Swarm.

codemac commented:
Well - let me know what changes you folks would like. I'm not sure I understand what #12931 is accomplishing in the "make the table easier for unix tools" goal of this pr.

tiborvass commented:
@calavera how do you want to do this? Is your WIP PR carrying this one?

calavera commented:
we'll, my pr is a different implementation in a more extensible way, but the final result is the same. I don't really have a problem with this, but if we're going to give flexibility to `docker ps`, I rather go all the way to my PR. As a first beneficial point, it already can add labels to the default ps output.

LK4D4 commented:
ping @docker/core-maintainers 
Should we close this or what?

duglin commented:
I'd prefer to have a smaller, more targeted solution, for now and wait until we have a proven need for the more flexible solution before we go for the other PR.

calavera commented:
#12931 has been updated to use go templates to be more clear on the format and follow the convention created with other commands. Although I haven't had time to update the description yet.

@duglin we already need a more flexible solution to support things like dynamic label outputs.

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "dockerps" git@github.com:codemac/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

aluzzardi commented:
What I'd like to see from this:

- Labels: As in, add columns based on a user specified labels
- Naming. Right now, columns are mapped based on hardcoded names, which is fine. I'd like to be able to rename them (as in, `CONTAINER ID` -> `ID`) and this is especially true for labels.


Use case:
```
docker ps --fields "c=ID,com.docker.swarm.id=Swarm ID,m"
```

/cc @vieux

codemac committed:
Support named labels with field=label syntax

* api/client/ps.go: named label support

Signed-off-by: Jeff Mickey <j@codemac.net>

codemac commented:
@aluzzardi see 59c4561b1ec33b35e4f9da548933ce05771a3ed0

codemac commented:
Well, I updated for naming the columns, but not any epic `--format`, `--template` or other work. I'd like to see a clear description of what is wanted from the steak-holders in this docker project who actually have a say. I've updated to mostly include what @aluzzardi asked for.

aluzzardi commented:
Thanks @codemac. The only missing requirement we (Swarm) have is to include container labels.

ping @docker/core-maintainers @vieux for feedback

calavera commented:
As @thaJeztah mentioned, and as we discussed in the maintainers meeting, any effort for custom formatting must:

1. Be more descriptive than single letters.
2. Follow Docker conventions about extra formatting by exposing a go template.

I really appreciate your PR, it made us aware of some improvements needed in this command. Please, refer to #12931 as it fulfills these two requirements and it also supports what the Swarm team needs.

duglin commented:
@codemac where are we with this?

icecrime commented:
@codemac Hey! Are you still available to rebase and pursue this PR? If that's not the case, maybe @calavera can you plese carry (meaning we keep your commits in and you still get credited for the work, but we take it on to completion). Thanks!

calavera commented:
@icecrime see https://github.com/docker/docker/pull/10255#issuecomment-103299280

I really don't care about having any credit in the commit, but we should not deviate from other commands that already support the `--format` option.

cpuguy83 commented:
@codemac Did you want to make the suggested changes?

icecrime commented:
Ping @docker/core-maintainers: if somebody wants to carry that, I think it's a good time.

arun-gupta commented:
Any ETA of when this will be available?

estesp commented:
I'm going to carry this--sorry for the delay as everything else has come up this week.  Of course, it will still require review and merge before it makes its way into master, so no prediction on an actual date.

estesp commented:
Carried in PR #14699 , closing.

thaJeztah commented:
Thanks @estesp!
